pkg/workspace/controllers/workspace_gc_finalizer.go

```diff
@@ -48,3 +59,101 @@ func (c *WorkspaceReconciler) garbageCollectWorkspace(ctx context.Context, wObj
+
+	// List PVCs in the same namespace with a label matching workspace name
+	pvcList := &corev1.PersistentVolumeClaimList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(wObj.Namespace),
+		client.MatchingLabels{kaitov1beta1.LabelWorkspaceName: wObj.Name},
+	}
+
+	if err := c.List(ctx, pvcList, listOpts...); err != nil {
+		klog.ErrorS(err, "failed to list PVCs for workspace", "workspace", klog.KObj(wObj))
+		return ctrl.Result{}, err
+	}
+
+	if len(pvcList.Items) == 0 {
+		klog.InfoS("No PVCs found for workspace", "workspace", klog.KObj(wObj))
+		return ctrl.Result{}, nil
+	}
+
+	// Process the found PVCs
+	var errs error
+	localpvCount := 0
+	for i := range pvcList.Items {
+		pvc := &pvcList.Items[i]
+		if pvc.Spec.StorageClassName != nil && *pvc.Spec.StorageClassName == consts.LocalNVMeStorageClass {
+			localpvCount++
+			if err := deleteOrphanedLocalPV(ctx, c.Client, pvc); err != nil {
+				errs = multierr.Append(errs, err)
+			}
+		}
+	}
+
+	if errs != nil {
+		return ctrl.Result{Requeue: true}, errs
+	}
+
+	if localpvCount > 0 {
+		return ctrl.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// When a node with local volumes gets removed from a cluster before deleting those volumes,
+// the PV and PVC objects may still exist. Force deleting objects in this case.
+// Ref: https://github.com/kubernetes-csi/external-provisioner?tab=readme-ov-file#deleting-local-volumes-after-a-node-failure-or-removal
+func deleteOrphanedLocalPV(ctx context.Context, c client.Client, pvc *corev1.PersistentVolumeClaim) error {
+	nodeName := pvc.Annotations["volume.kubernetes.io/selected-node"]
+	// If pvc is not bound, there's nothing to do
+	if nodeName == "" || pvc.Status.Phase != corev1.ClaimBound {
+		return nil
+	}
+
+	// Check if the node exists
+	node := &corev1.Node{}
+	err := c.Get(ctx, client.ObjectKey{Name: nodeName}, node)
+	if err == nil {
+		// Node exists, waiting for csi driver to clean up
+		return nil
+	}
+	if client.IgnoreNotFound(err) != nil {
+		klog.ErrorS(err, "Failed to get node", "node", nodeName, "pvc", klog.KObj(pvc))
+		return err
+	}
+
+	// Node does not exist, force delete pv
+	klog.InfoS("Node does not exist, deleting orphaned PVC", "pvc", klog.KObj(pvc), "node", nodeName)
+	pvName := pvc.Spec.VolumeName
+	if pvName == "" {
+		klog.Warningf("PVC %s does not have a bound PV, nothing to do", klog.KObj(pvc))
+		return nil
+	}
+
+	pv := &corev1.PersistentVolume{}
+	if err := c.Get(ctx, client.ObjectKey{Name: pvName}, pv); err == nil {
+		// PV exists, set finalizers to nil
+		pv.Finalizers = nil
+		if err := c.Update(ctx, pv); err != nil {
+			klog.ErrorS(err, "Failed to update PV finalizers", "pv", klog.KObj(pv))
+			return err
+		}
+		klog.InfoS("Removed finalizers from PV", "pv", klog.KObj(pv))
+	} else if client.IgnoreNotFound(err) != nil {
+		klog.ErrorS(err, "Failed to get PV", "pv", pvName)
+		return err
+	}
```

Comment:
Can we simply delete the PVC and let the CSI driver controller clean up the PV for us?

Reply:
No, it can't. The main reason to add this logic is that

> When a node with local volumes gets removed from a cluster before deleting those volumes, the PV and PVC objects may still exist. normal deletion of the volume and thus deletion of the PV is not possible anymore because the CSI driver instance on the node is not available

It's a special local pv csi driver which deploy in daemonset. We need to clean them manually in edge cases.
https://github.com/kubernetes-csi/external-provisioner?tab=readme-ov-file#deleting-local-volumes-after-a-node-failure-or-removal